File: SDK/0.0.5_android/android_aar/src/main/java/com/wenet/asr/WavReader.java
package com.wenet.asr;

import java.io.ByteArrayOutputStream;
import java.io.EOFException;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

final class WavReader {
    private WavReader() {}

    static WavData readPcm16(File file) throws IOException {
        try (InputStream in = new FileInputStream(file)) {
            return readPcm16(in);
        }
    }

    static WavData readPcm16(InputStream in) throws IOException {
        if (!"RIFF".equals(readTag(in))) {
            throw new IOException("missing RIFF header");
        }
        skipFully(in, 4);
        if (!"WAVE".equals(readTag(in))) {
            throw new IOException("missing WAVE header");
        }

        int audioFormat = 0;
        int channels = 0;
        int sampleRate = 0;
        int bitsPerSample = 0;
        byte[] pcmBytes = null;

        while (audioFormat == 0 || pcmBytes == null) {
            String tag = readTagOrNull(in);
            if (tag == null) {
                break;
            }
            int size = readU32(in);
            if ("fmt ".equals(tag)) {
                audioFormat = readU16(in);
                channels = readU16(in);
                sampleRate = readU32(in);
                skipFully(in, 6);
                bitsPerSample = readU16(in);
                if (size > 16) {
                    skipFully(in, size - 16);
                }
            } else if ("data".equals(tag)) {
                pcmBytes = readBytes(in, size);
            } else {
                skipFully(in, size);
            }
            if ((size & 1) == 1) {
                skipFully(in, 1);
            }
        }

        if (audioFormat != 1 || bitsPerSample != 16 || channels <= 0) {
            throw new IOException("only 16-bit PCM WAV is supported");
        }
        if (pcmBytes == null) {
            throw new IOException("missing WAV data chunk");
        }

        int totalSamples = pcmBytes.length / 2;
        int monoSamples = totalSamples / channels;
        short[] samples = new short[monoSamples];
        for (int i = 0; i < monoSamples; ++i) {
            int mixed = 0;
            for (int c = 0; c < channels; ++c) {
                int byteIndex = (i * channels + c) * 2;
                int lo = pcmBytes[byteIndex] & 0xff;
                int hi = pcmBytes[byteIndex + 1] << 8;
                mixed += (short) (lo | hi);
            }
            samples[i] = (short) (mixed / channels);
        }
        return new WavData(sampleRate, samples);
    }

    private static String readTag(InputStream in) throws IOException {
        byte[] bytes = readBytes(in, 4);
        return new String(bytes, "US-ASCII");
    }

    private static String readTagOrNull(InputStream in) throws IOException {
        byte[] bytes = new byte[4];
        int offset = 0;
        while (offset < bytes.length) {
            int read = in.read(bytes, offset, bytes.length - offset);
            if (read < 0) {
                return offset == 0 ? null : failEof();
            }
            offset += read;
        }
        return new String(bytes, "US-ASCII");
    }

    private static String failEof() throws EOFException {
        throw new EOFException("unexpected end of WAV file");
    }

    private static int readU16(InputStream in) throws IOException {
        byte[] b = readBytes(in, 2);
        return (b[0] & 0xff) | ((b[1] & 0xff) << 8);
    }

    private static int readU32(InputStream in) throws IOException {
        byte[] b = readBytes(in, 4);
        long value = (b[0] & 0xffL)
                | ((b[1] & 0xffL) << 8)
                | ((b[2] & 0xffL) << 16)
                | ((b[3] & 0xffL) << 24);
        if (value > Integer.MAX_VALUE) {
            throw new IOException("WAV chunk is too large");
        }
        return (int) value;
    }

    private static byte[] readBytes(InputStream in, int size) throws IOException {
        byte[] bytes = new byte[size];
        int offset = 0;
        while (offset < size) {
            int read = in.read(bytes, offset, size - offset);
            if (read < 0) {
                throw new EOFException("unexpected end of WAV file");
            }
            offset += read;
        }
        return bytes;
    }

    private static void skipFully(InputStream in, int size) throws IOException {
        int remaining = size;
        byte[] buffer = null;
        while (remaining > 0) {
            long skipped = in.skip(remaining);
            if (skipped > 0) {
                remaining -= (int) skipped;
                continue;
            }
            if (buffer == null) {
                buffer = new byte[4096];
            }
            int read = in.read(buffer, 0, Math.min(buffer.length, remaining));
            if (read < 0) {
                throw new EOFException("unexpected end of WAV file");
            }
            remaining -= read;
        }
    }
}

File: SDK/0.0.6/src/audio/wav_reader.cc
#include "audio/wav_reader.h"

#include <algorithm>
#include <cmath>
#include <cstring>
#include <fstream>
#include <limits>
#include <string>

namespace asr_sdk::internal {
namespace {

uint32_t ReadU32(std::ifstream* in) {
  unsigned char b[4] = {0, 0, 0, 0};
  in->read(reinterpret_cast<char*>(b), 4);
  return static_cast<uint32_t>(b[0]) | (static_cast<uint32_t>(b[1]) << 8) |
         (static_cast<uint32_t>(b[2]) << 16) |
         (static_cast<uint32_t>(b[3]) << 24);
}

uint16_t ReadU16(std::ifstream* in) {
  unsigned char b[2] = {0, 0};
  in->read(reinterpret_cast<char*>(b), 2);
  return static_cast<uint16_t>(b[0]) | (static_cast<uint16_t>(b[1]) << 8);
}

std::string ReadTag(std::ifstream* in) {
  char tag[4] = {0, 0, 0, 0};
  in->read(tag, 4);
  return std::string(tag, 4);
}

}  // namespace

Status ReadWavFile(const std::filesystem::path& path, WavData* data) {
  if (data == nullptr) {
    return Status::InvalidArgument("wav output is null");
  }
  std::ifstream in(path, std::ios::binary);
  if (!in) {
    return Status::NotFound("failed to open wav: " + path.string());
  }
  if (ReadTag(&in) != "RIFF") {
    return Status::InvalidArgument("wav is missing RIFF header: " +
                                   path.string());
  }
  (void)ReadU32(&in);
  if (ReadTag(&in) != "WAVE") {
    return Status::InvalidArgument("wav is missing WAVE header: " +
                                   path.string());
  }

  uint16_t audio_format = 0;
  uint16_t channels = 0;
  uint32_t sample_rate = 0;
  uint16_t bits_per_sample = 0;
  std::vector<char> pcm_bytes;

  while (in && (!audio_format || pcm_bytes.empty())) {
    const std::string tag = ReadTag(&in);
    if (!in) {
      break;
    }
    const uint32_t size = ReadU32(&in);
    if (tag == "fmt ") {
      audio_format = ReadU16(&in);
      channels = ReadU16(&in);
      sample_rate = ReadU32(&in);
      (void)ReadU32(&in);
      (void)ReadU16(&in);
      bits_per_sample = ReadU16(&in);
      if (size > 16) {
        in.seekg(static_cast<std::streamoff>(size - 16), std::ios::cur);
      }
    } else if (tag == "data") {
      pcm_bytes.resize(size);
      in.read(pcm_bytes.data(), static_cast<std::streamsize>(size));
    } else {
      in.seekg(static_cast<std::streamoff>(size), std::ios::cur);
    }
    if (size % 2 == 1) {
      in.seekg(1, std::ios::cur);
    }
  }

  const bool pcm16 = audio_format == 1 && bits_per_sample == 16;
  const bool float32 = audio_format == 3 && bits_per_sample == 32;
  if ((!pcm16 && !float32) || channels == 0) {
    return Status::InvalidArgument(
        "only 16-bit PCM or 32-bit IEEE float WAV files are supported by this tool");
  }
  data->sample_rate = static_cast<int>(sample_rate);
  data->num_channels = static_cast<int>(channels);
  data->samples.clear();

  const size_t bytes_per_sample = pcm16 ? sizeof(int16_t) : sizeof(float);
  if (pcm_bytes.size() % (bytes_per_sample * channels) != 0) {
    return Status::InvalidArgument("wav data chunk is not sample aligned: " +
                                   path.string());
  }
  const size_t total_samples = pcm_bytes.size() / bytes_per_sample;
  data->samples.reserve(total_samples / channels);
  for (size_t i = 0; i + channels <= total_samples; i += channels) {
    double mix = 0.0;
    for (uint16_t c = 0; c < channels; ++c) {
      const size_t byte = (i + c) * bytes_per_sample;
      if (pcm16) {
        const auto lo = static_cast<unsigned char>(pcm_bytes[byte]);
        const auto hi = static_cast<unsigned char>(pcm_bytes[byte + 1]);
        const int16_t sample =
            static_cast<int16_t>(lo | static_cast<uint16_t>(hi << 8));
        mix += sample;
      } else {
        uint32_t bits = static_cast<unsigned char>(pcm_bytes[byte]) |
                        (static_cast<uint32_t>(
                             static_cast<unsigned char>(pcm_bytes[byte + 1]))
                         << 8) |
                        (static_cast<uint32_t>(
                             static_cast<unsigned char>(pcm_bytes[byte + 2]))
                         << 16) |
                        (static_cast<uint32_t>(
                             static_cast<unsigned char>(pcm_bytes[byte + 3]))
                         << 24);
        float sample = 0.0f;
        std::memcpy(&sample, &bits, sizeof(sample));
        if (std::isfinite(sample)) {
          mix += std::clamp(static_cast<double>(sample), -1.0, 1.0) *
                 32767.0;
        }
      }
    }
    const double mono = mix / channels;
    const double clipped = std::clamp(
        mono, static_cast<double>(std::numeric_limits<int16_t>::min()),
        static_cast<double>(std::numeric_limits<int16_t>::max()));
    data->samples.push_back(static_cast<int16_t>(std::lround(clipped)));
  }
  return Status::Ok();
}

}  // namespace asr_sdk::internal
